src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -2873,6 +2873,322 @@ static void Main()
+    {
+        var source = """
+            using System.Runtime.InteropServices;
+            class C
+            {
+                static void M([Optional, DefaultParameterValue(1)] ref readonly int i) => System.Console.Write(i);
+                static void Main()
+                {
+                    int x = 2;
+                    M();
+                    M(x);
+                    M(ref x);
+                    M(in x);
+                }
+                static void M2() => M();
+            }
+            """;
+        var verifier = CompileAndVerify(source, expectedOutput: "1222").VerifyDiagnostics(
+            // (4,30): warning CS9521: A default value is specified for 'ref readonly' parameter 'i', but 'ref readonly' should be used only for references. Consider declaring the parameter as 'in'.
+            //     static void M([Optional, DefaultParameterValue(1)] ref readonly int i) => System.Console.Write(i);
+            Diagnostic(ErrorCode.WRN_RefReadonlyParameterDefaultValue, "DefaultParameterValue(1)").WithArguments("i").WithLocation(4, 30),
+            // (9,11): warning CS9503: Argument 1 should be passed with 'ref' or 'in' keyword
+            //         M(x);
+            Diagnostic(ErrorCode.WRN_ArgExpectedRefOrIn, "x").WithArguments("1").WithLocation(9, 11));
+        verifier.VerifyIL("C.M2", """
+            {
+              // Code size       10 (0xa)
+              .maxstack  1
+              .locals init (int V_0)
+              IL_0000:  ldc.i4.1
+              IL_0001:  stloc.0
+              IL_0002:  ldloca.s   V_0
+              IL_0004:  call       "void C.M(ref readonly int)"
+              IL_0009:  ret
+            }
+            """);
+    }
+
+    [Fact]
+    public void DefaultParameterValue_AttributeAndEqualsValue()
+    {
+        var source = """
+            using System.Runtime.InteropServices;
+            class C
+            {
+                static void M([DefaultParameterValue(1)] ref readonly int i = 1) => throw null;
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (4,20): error CS1745: Cannot specify default parameter value in conjunction with DefaultParameterAttribute or OptionalAttribute
+            //     static void M([DefaultParameterValue(1)] ref readonly int i = 1) => throw null;
+            Diagnostic(ErrorCode.ERR_DefaultValueUsedWithAttributes, "DefaultParameterValue").WithLocation(4, 20),
+            // (4,67): warning CS9521: A default value is specified for 'ref readonly' parameter 'i', but 'ref readonly' should be used only for references. Consider declaring the parameter as 'in'.
+            //     static void M([DefaultParameterValue(1)] ref readonly int i = 1) => throw null;
+            Diagnostic(ErrorCode.WRN_RefReadonlyParameterDefaultValue, "1").WithArguments("i").WithLocation(4, 67),
+            // (4,67): error CS8017: The parameter has multiple distinct default values.
```

Comment:
>// (4,67): error CS8017: The parameter has multiple distinct default values. [](http://example.com/codeflow?start=12&length=76)

This is probably unrelated to the changes, but why is this error reported?

Reply:
> I guess the latter should not report if any of the values is bad

Probably. Could you open an issue for this?

Reply:
https://github.com/dotnet/roslyn/issues/69078